### PR TITLE
Update rules UI to include optional user scope

### PR DESCRIPTION
**Pull request by akrousset**
Fixes firezone/product#394 

Refs #423

### CODE REVIEW

**Comment by conectado** [comment]
I think it'd be a good idea to add a list of rules in the user view, and maybe the option to add/remove rules from there?

**Comment by jamilbk** [comment]
This looks good so far! Need to think more about @conectado ‘s comment re: rules table. I think what could work for now is showing a simple rule count in the user details table when an admin is viewing a user.

This could link to the rules page

**Comment by jamilbk** [comment]
Also — would you be able to add some tests to `rules_test.exs` to test this new field?

File: apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex [hunk 1]
```diff
@@ -40,6 +49,9 @@
                   <%= rule.destination %>
                 </dd>
               </td>
+              <td class="has-text-left">
+                <%= get_scoped_user(rule.user_id) %>
```
Comment (princemaple) [inline apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex]: This looks like an N+1? If we are already loading all users, we might as well build a map out of it and use it here.

**Review by princemaple — CHANGES_REQUESTED** [review]

File: apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex (hunk 1 above)
Comment (jamilbk) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex]: 👍

File: apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex [hunk 2]
```diff
@@ -40,6 +49,9 @@
                   <%= rule.destination %>
                 </dd>
               </td>
+              <td class="has-text-left">
+                <%= @users[rule.user_id] %>
```
Comment (princemaple) [inline apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex:53]: Will there be rule.user_id == current_user_id?

Comment (akrousset) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex:53]: The current user id is filtered out before building the map:

https://github.com/firezone/firezone/blob/21394ca539b147b088d41914ff742e846b3fed26/apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.ex#L74-L77

Comment (princemaple) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex:53]: I thought we'd want the current user in the map. Maybe I didn't get what this is trying to achieve.

Comment (akrousset) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex:53]: Yeah, hmmm 🤔 

The current user would be the admin, so I was assuming as an admin you would only scope rules to your users and not yourself?

Comment (princemaple) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex:53]: What if the admin wants some special sauce for themself :)

Also an admin could add a rule for another admin, right?
In that case, you wouldn't want to mistake a rule that only applies to you as global.

Comment (akrousset) [reply apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex:53]: Right, that makes more sense. Thanks for clarifying for me!

**Review by princemaple — APPROVED** [review]

**Review by conectado — COMMENTED** [review]
Looks good! Just waiting for the changes described [here](https://github.com/firezone/firezone/pull/764#issuecomment-1168021148) to approve

File: apps/fz_http/lib/fz_http_web/templates/shared/user_details.html.heex [hunk 3]
```diff
@@ -45,9 +45,11 @@
       <td><%= FzHttp.Devices.count(@user.id) %></td>
     </tr>
 
-    <tr>
-      <td><strong>Number of Rules</strong></td>
-      <td><a href={"#{@rules_path}"}><%= FzHttp.Rules.count(@user.id) %></a></td>
-    </tr>
+    <%= if @rules_path do %>
```
Comment (akrousset) [inline apps/fz_http/lib/fz_http_web/templates/shared/user_details.html.heex:48]: I'm conditionally showing this here, I figured just admin users would want to see the number of rules.

**Review by conectado — APPROVED** [review]
Looks Great!

File: apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex [hunk 4]
```diff
@@ -15,6 +15,14 @@
               class: "input #{input_error_class(f, :destination)}",
               placeholder: "IPv4/6 CIDR range or address" %>
         </div>
+        <div class="control">
+          <div class="select">
+            <%= select f,
+                :user_id,
+                Enum.map(@users, fn {id, email} -> {email, id} end),
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/rule_live/rule_list_component.html.heex]: suggestion: move this to the `assigns` or a helper function to keep logic out of the view

File: apps/fz_http/lib/fz_http_web/live/setting_live/account_live.ex [hunk 5]
```diff
@@ -18,6 +18,7 @@ defmodule FzHttpWeb.SettingLive.Account do
      |> assign(:changeset, Users.change_user(socket.assigns.current_user))
      |> assign(:methods, MFA.list_methods(socket.assigns.current_user))
      |> assign(:page_title, "Account Settings")
+     |> assign(:rules_path, Routes.rule_index_path(FzHttpWeb.Endpoint, :index))
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/setting_live/account_live.ex]: Replace `FzHttpWeb.Endpoint` with `socket` (convention we've been using)

File: apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex [hunk 6]
```diff
@@ -20,7 +20,8 @@ defmodule FzHttpWeb.UserLive.Show do
      |> assign(:device_config, socket.assigns[:device_config])
      |> assign(:connections, connections)
      |> assign(:user, user)
-     |> assign(:page_title, "Users")}
+     |> assign(:page_title, "Users")
+     |> assign(:rules_path, Routes.rule_index_path(FzHttpWeb.Endpoint, :index))}
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/user_live/show_live.ex]: same re: `socket`

File: apps/fz_http/lib/fz_http_web/templates/shared/user_details.html.heex (hunk 3 above)
Comment (jamilbk) [reply apps/fz_http/lib/fz_http_web/templates/shared/user_details.html.heex:48]: 👍

**Review by jamilbk — CHANGES_REQUESTED** [review]
LGTM, just some style cleanup

**Comment by jamilbk** [comment]
@akrousset One last thing on this -- would you be able to update the corresponding docs for this change?

Check docs/docs/user-guides/firewall-rules.md

Probably don't have to worry about the screenshot update, but probably should add some explanation around how this works.

**Comment by akrousset** [comment]
I've added an issue to update the image firezone/product#408

**Pull request merged**